File: backend/services/hone/domain/reading.go
package domain

import (
	"context"
	"time"

	"github.com/google/uuid"
)

// Reading sub-context (Wave 4 of docs/feature/english.md). Three flat
// entities — materials, sessions, vocab queue — owned by Hone because
// the user-facing surface is Hone's Reading-модуль (hotkey R), not
// the web. Engineering-only tables (ratings/elo/tasks) are not
// touched; English content lives free-form alongside Hone notes.

// ReadingSourceKind enumerates how a material was ingested. Mirrors
// the CHECK constraint on hone_reading_materials.source_kind. The
// frontend stays free to pick the input flow (paste / URL fetch /
// file upload) without forcing a schema migration per channel.
type ReadingSourceKind string

const (
	ReadingSourcePaste ReadingSourceKind = "paste"
	ReadingSourceURL   ReadingSourceKind = "url"
	ReadingSourcePDF   ReadingSourceKind = "pdf"
	ReadingSourceEPUB  ReadingSourceKind = "epub"
)

// IsValid keeps switches downstream exhaustive.
func (k ReadingSourceKind) IsValid() bool {
	switch k {
	case ReadingSourcePaste, ReadingSourceURL, ReadingSourcePDF, ReadingSourceEPUB:
		return true
	}
	return false
}

// ReadingMaterial mirrors a row in hone_reading_materials. BodyMD is
// the full content; total_chars is precomputed at insert time so
// progress-bar widgets don't have to recount.
type ReadingMaterial struct {
	ID         uuid.UUID
	UserID     uuid.UUID
	SourceKind ReadingSourceKind
	SourceURL  string // empty for SourceKind == paste
	Title      string
	BodyMD     string
	TotalChars int
	ArchivedAt *time.Time // soft-delete; nil = active
	CreatedAt  time.Time
	UpdatedAt  time.Time
}

// ReadingSession is one «sit» of reading. Started when the user
// opens the material; ended_at stamped when they close it or the
// pomodoro timer expires.
//
// AISummaryScore + SummaryMD are populated by the post-chapter AI
// summary-check (Wave 4.3 — outside this slice). Both nil/empty
// when the session ends without a summary attempt.
type ReadingSession struct {
	ID             uuid.UUID
	UserID         uuid.UUID
	MaterialID     uuid.UUID
	CharsRead      int
	CharsTotal     int
	StartedAt      time.Time
	EndedAt        *time.Time
	AISummaryScore *int
	SummaryMD      string
}

// VocabEntry is one row in hone_vocab_queue. Box (0..5) drives the
// SRS scheduler — review intervals scale by box level. NextReviewAt
// is the «when to surface this card next» key; LearnedAt non-nil
// means the user has graduated this word and it's no longer in the
// active queue (kept for analytics).
type VocabEntry struct {
	UserID         uuid.UUID
	Word           string
	Translation    string
	ContextMD      string
	SourceMaterial *uuid.UUID
	Box            int
	NextReviewAt   time.Time
	ReviewedCount  int
	LearnedAt      *time.Time
	CreatedAt      time.Time
}

// ReadingRepo is the persistence surface. One interface per sub-
// context (materials / sessions / vocab) would force three injects
// at every callsite — Hone's other sub-contexts use one fat repo
// per area too (NoteRepo, FocusRepo, etc.), so we follow suit.
type ReadingRepo interface {
	// CreateMaterial persists a new material. Caller fills Title,
	// SourceKind, BodyMD; the repo stamps ID + timestamps and
	// computes TotalChars from BodyMD length.
	CreateMaterial(ctx context.Context, m ReadingMaterial) (ReadingMaterial, error)

	// GetMaterial loads a single material; ErrNotFound when absent
	// OR when the requested user_id doesn't own it (cross-user leak
	// protection — same convention as other Hone repos).
	GetMaterial(ctx context.Context, userID, materialID uuid.UUID) (ReadingMaterial, error)

	// ListMaterials returns the user's active materials (ArchivedAt
	// IS NULL), most-recent first. limit caps the result.
	ListMaterials(ctx context.Context, userID uuid.UUID, limit int) ([]ReadingMaterial, error)

	// ArchiveMaterial soft-deletes a material. Returns ErrNotFound
	// if the row doesn't exist or belongs to another user.
	ArchiveMaterial(ctx context.Context, userID, materialID uuid.UUID, now time.Time) error

	// StartSession creates a hone_reading_sessions row and returns
	// it (with ID stamped). chars_total is captured from the
	// material at start-time so a later body edit doesn't lie about
	// «I read 90% of the article».
	StartSession(ctx context.Context, userID, materialID uuid.UUID) (ReadingSession, error)

	// EndSession stamps ended_at and writes chars_read + summary_md
	// (when the user submitted a summary). ai_summary_score stays
	// nil here — the AI grader is a separate flow (Wave 4.3).
	EndSession(ctx context.Context, userID, sessionID uuid.UUID, charsRead int, summaryMD string, now time.Time) error

	// GetSession loads a single session row. Used by the post-end
	// grader path (read-modify-write of ai_summary_score) and by
	// future analytics callers. ErrNotFound on cross-user probe.
	GetSession(ctx context.Context, userID, sessionID uuid.UUID) (ReadingSession, error)

	// SetAISummaryScore persists the AI-graded score (0..100) onto a
	// completed session. Idempotent — re-grading the same session
	// overwrites the previous score; we don't keep a history.
	SetAISummaryScore(ctx context.Context, userID, sessionID uuid.UUID, score int) error

	// ListVocabDue returns vocab entries due today (next_review_at
	// <= now), ordered earliest-first. Drives the daily 5-minute
	// review widget. Excludes learned_at IS NOT NULL.
	ListVocabDue(ctx context.Context, userID uuid.UUID, now time.Time, limit int) ([]VocabEntry, error)

	// UpsertVocab adds a word to the queue OR refreshes its
	// context_md if it already exists. Idempotent — clicking the
	// same word twice in the same chapter doesn't reset the box.
	UpsertVocab(ctx context.Context, e VocabEntry) (VocabEntry, error)

	// AdvanceVocab moves a vocab card forward in the SRS algorithm.
	// `correct` distinguishes graduating (box+1) from regressing
	// (box=0). LearnedAt is stamped when box reaches 5.
	AdvanceVocab(ctx context.Context, userID uuid.UUID, word string, correct bool, now time.Time) (VocabEntry, error)

	// ListVocabBySourceMaterial — Wave 4.2 reverse cross-link. Returns
	// every vocab entry whose `source_material` points to materialID,
	// scoped to the user. Reader UI shows «vocab you've saved from this
	// material» sidebar; without this method that surface would require
	// the client to fetch the entire vocab queue and filter.
	// limit caps the result; 0 → server default (50).
	ListVocabBySourceMaterial(ctx context.Context, userID, materialID uuid.UUID, limit int) ([]VocabEntry, error)
}

// SummaryGrader scores a user-written summary against the source
// material on a 0..100 scale. Wave 4.3 «AI summary check»: after
// the user finishes a chapter and submits their summary, the grader
// compares it to the chapter body and returns coverage + accuracy.
//
// Implementations are LLM-backed; an explicit no-op fallback returns
// ErrLLMUnavailable when the chain isn't configured. Use cases that
// call this MUST treat both ErrLLMUnavailable and any other error as
// non-fatal — the session was already persisted by EndSession, the
// grade is best-effort.
type SummaryGrader interface {
	GradeSummary(ctx context.Context, in GradeSummaryInput) (int, error)
}

// GradeSummaryInput carries everything the LLM needs to grade. Title
// is included so the prompt can scope its judgement to the chapter,
// not the whole book/article series.
type GradeSummaryInput struct {
	Title   string
	BodyMD  string
	Summary string
}

File: hone/src/renderer/src/api/reading.ts
// api/reading.ts — Reading-модуль API client (Wave 4 of english.md).
// Library + reader sessions + Leitner SRS vocab. Same pattern as api/hone.ts —
// module-private Connect client, named async wrappers return POJOs так чтобы
// UI never sees proto-message classes.
import { createPromiseClient } from '@connectrpc/connect';
import { HoneService } from '@generated/pb/druz9/v1/hone_connect';

import { transport } from './transport';

// ─── Domain-shaped POJOs ───────────────────────────────────────────────────

export type ReadingSourceKind = 'paste' | 'url' | 'pdf' | 'epub' | 'book';

export interface ReadingMaterial {
  id: string;
  sourceKind: ReadingSourceKind;
  sourceUrl: string;
  title: string;
  bodyMd: string; // empty in list responses
  totalChars: number;
  archivedAt: Date | null;
  createdAt: Date | null;
  updatedAt: Date | null;
  // Book-only progress (Sergey 2026-05-03). null когда не задано.
  bookChapter: number | null;
  bookTotalChapters: number | null;
}

export interface ReadingSession {
  id: string;
  materialId: string;
  charsRead: number;
  charsTotal: number;
  startedAt: Date | null;
  endedAt: Date | null;
  aiSummaryScore: number | null; // null when not yet graded
  summaryMd: string;
}

export interface VocabEntry {
  word: string;
  translation: string;
  contextMd: string;
  sourceMaterial: string | null;
  box: number; // 0..5; 5 = graduated
  nextReviewAt: Date | null;
  reviewedCount: number;
  learnedAt: Date | null;
  createdAt: Date | null;
}

// ─── Internals ─────────────────────────────────────────────────────────────

const client = createPromiseClient(HoneService, transport);

function protoTs(ts: { seconds: bigint; nanos: number } | undefined): Date | null {
  if (!ts) return null;
  const ms = Number(ts.seconds) * 1000 + Math.floor(ts.nanos / 1_000_000);
  if (ms <= 0) return null; // backend leaves zero-stamp on "not set" fields
  return new Date(ms);
}

function normalizeSourceKind(k: string): ReadingSourceKind {
  switch (k) {
    case 'paste':
    case 'url':
    case 'pdf':
    case 'epub':
    case 'book':
      return k;
    default:
      return 'paste';
  }
}

type ProtoMaterial = {
  id: string;
  sourceKind: string;
  sourceUrl: string;
  title: string;
  bodyMd: string;
  totalChars: number;
  archivedAt?: { seconds: bigint; nanos: number };
  createdAt?: { seconds: bigint; nanos: number };
  updatedAt?: { seconds: bigint; nanos: number };
  bookChapter?: number;
  hasBookChapter?: boolean;
  bookTotalChapters?: number;
  hasBookTotal?: boolean;
};

function unwrapMaterial(m: ProtoMaterial): ReadingMaterial {
  return {
    id: m.id,
    sourceKind: normalizeSourceKind(m.sourceKind),
    sourceUrl: m.sourceUrl,
    title: m.title,
    bodyMd: m.bodyMd,
    totalChars: m.totalChars,
    archivedAt: protoTs(m.archivedAt),
    createdAt: protoTs(m.createdAt),
    updatedAt: protoTs(m.updatedAt),
    bookChapter: m.hasBookChapter ? (m.bookChapter ?? 0) : null,
    bookTotalChapters: m.hasBookTotal ? (m.bookTotalChapters ?? 0) : null,
  };
}

type ProtoSession = {
  id: string;
  materialId: string;
  charsRead: number;
  charsTotal: number;
  startedAt?: { seconds: bigint; nanos: number };
  endedAt?: { seconds: bigint; nanos: number };
  aiSummaryScore: number;
  hasScore: boolean;
  summaryMd: string;
};

function unwrapSession(s: ProtoSession): ReadingSession {
  return {
    id: s.id,
    materialId: s.materialId,
    charsRead: s.charsRead,
    charsTotal: s.charsTotal,
    startedAt: protoTs(s.startedAt),
    endedAt: protoTs(s.endedAt),
    aiSummaryScore: s.hasScore ? s.aiSummaryScore : null,
    summaryMd: s.summaryMd,
  };
}

type ProtoVocab = {
  word: string;
  translation: string;
  contextMd: string;
  sourceMaterial: string;
  box: number;
  nextReviewAt?: { seconds: bigint; nanos: number };
  reviewedCount: number;
  learnedAt?: { seconds: bigint; nanos: number };
  createdAt?: { seconds: bigint; nanos: number };
};

function unwrapVocab(v: ProtoVocab): VocabEntry {
  return {
    word: v.word,
    translation: v.translation,
    contextMd: v.contextMd,
    sourceMaterial: v.sourceMaterial.length > 0 ? v.sourceMaterial : null,
    box: v.box,
    nextReviewAt: protoTs(v.nextReviewAt),
    reviewedCount: v.reviewedCount,
    learnedAt: protoTs(v.learnedAt),
    createdAt: protoTs(v.createdAt),
  };
}

// ─── Materials ─────────────────────────────────────────────────────────────

export async function listReadingMaterials(limit = 100): Promise<ReadingMaterial[]> {
  const resp = await client.listReadingMaterials({ limit });
  return resp.items.map((m) => unwrapMaterial(m as unknown as ProtoMaterial));
}

export async function getReadingMaterial(id: string): Promise<ReadingMaterial> {
  const resp = await client.getReadingMaterial({ id });
  return unwrapMaterial(resp as unknown as ProtoMaterial);
}

export async function addReadingMaterial(args: {
  sourceKind: ReadingSourceKind;
  title: string;
  bodyMd: string;
  sourceUrl?: string;
  bookChapter?: number;
  bookTotalChapters?: number;
}): Promise<ReadingMaterial> {
  const resp = await client.addReadingMaterial({
    sourceKind: args.sourceKind,
    title: args.title,
    bodyMd: args.bodyMd,
    sourceUrl: args.sourceUrl ?? '',
    bookChapter: args.bookChapter ?? 0,
    hasBookChapter: typeof args.bookChapter === 'number',
    bookTotalChapters: args.bookTotalChapters ?? 0,
    hasBookTotal: typeof args.bookTotalChapters === 'number',
  });
  return unwrapMaterial(resp as unknown as ProtoMaterial);
}

export async function updateBookProgress(args: {
  id: string;
  bookChapter?: number;
  bookTotalChapters?: number;
}): Promise<ReadingMaterial> {
  const resp = await client.updateBookProgress({
    id: args.id,
    bookChapter: args.bookChapter ?? 0,
    hasBookChapter: typeof args.bookChapter === 'number',
    bookTotalChapters: args.bookTotalChapters ?? 0,
    hasBookTotal: typeof args.bookTotalChapters === 'number',
  });
  return unwrapMaterial(resp as unknown as ProtoMaterial);
}

export async function archiveReadingMaterial(id: string): Promise<void> {
  await client.archiveReadingMaterial({ id });
}

// ─── Sessions ─────────────────────────────────────────────────────────────

export async function startReadingSession(materialId: string): Promise<ReadingSession> {
  const resp = await client.startReadingSession({ materialId });
  return unwrapSession(resp as unknown as ProtoSession);
}

// endReadingSession returns the closed session, including ai_summary_score
// when the LLM grader finished within the request budget. If the grader
// is offline or timed out, `aiSummaryScore` is null and the caller should
// just treat the session as done-without-score.
export async function endReadingSession(args: {
  sessionId: string;
  charsRead: number;
  summaryMd?: string;
}): Promise<ReadingSession> {
  const resp = await client.endReadingSession({
    sessionId: args.sessionId,
    charsRead: args.charsRead,
    summaryMd: args.summaryMd ?? '',
  });
  if (!resp.session) {
    // Defensive: an old server (pre-4.3) returned an empty response. Synth
    // a stub so the caller's typed flow keeps working — they'll just see
    // no score, which matches the pre-4.3 UX.
    return {
      id: args.sessionId,
      materialId: '',
      charsRead: args.charsRead,
      charsTotal: 0,
      startedAt: null,
      endedAt: new Date(),
      aiSummaryScore: null,
      summaryMd: args.summaryMd ?? '',
    };
  }
  return unwrapSession(resp.session as unknown as ProtoSession);
}

// ─── Vocab queue ──────────────────────────────────────────────────────────

export async function addVocab(args: {
  word: string;
  translation?: string;
  contextMd?: string;
  sourceMaterial?: string;
}): Promise<VocabEntry> {
  const resp = await client.addVocab({
    word: args.word,
    translation: args.translation ?? '',
    contextMd: args.contextMd ?? '',
    sourceMaterial: args.sourceMaterial ?? '',
  });
  return unwrapVocab(resp as unknown as ProtoVocab);
}

export async function reviewVocab(word: string, correct: boolean): Promise<VocabEntry> {
  const resp = await client.reviewVocab({ word, correct });
  return unwrapVocab(resp as unknown as ProtoVocab);
}

/** Wave 4.2 — vocab entries saved from a specific source material.
 *  Powers the «words you've saved here» reader sidebar. */
export async function listVocabBySourceMaterial(materialId: string, limit = 50): Promise<VocabEntry[]> {
  const resp = await client.listVocabBySourceMaterial({ materialId, limit });
  return resp.items.map((v) => unwrapVocab(v as unknown as ProtoVocab));
}

export async function listVocabDue(limit = 20): Promise<VocabEntry[]> {
  const resp = await client.listVocabDue({ limit });
  return resp.items.map((v) => unwrapVocab(v as unknown as ProtoVocab));
}
